function_migrations.go

```diff
@@ -0,0 +1,92 @@
+package function
+
+import (
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+)
+
+// Migrate applies any necessary migrations, returning a new migrated
+// version of the Function.  It is the caller's responsibility to
+// .Write() the Function to persist to disk.
+func (f Function) Migrate() (migrated Function, err error) {
+	// Return immediately if the Function indicates it has already been
+	// migrated.
+	if f.Migrated() {
+		return f, nil
```

Comment:
Would it be worth testing this scenario with a `func.yaml` that already has a version?

Reply:
Yes this is probably a good idea.  After considering it a bit, this really boils down to .Migrate needing to be idempotent.
This boiling down can be bubbled up to asserting that .Write should also be idempotent.  Therefore I added a test which makes this expectation explicit, and it flushed out some needed tweaks to Function initial values in the process.  Thanks for the suggestion.